RHPAM-2727 - Stunner - it is not possible to open process without simulation extensions

hasys opened the pull request:
Hi @romartin, @LuboTerifaj with this fix I was able to load 3 from 4 diagrams provided to me in RHPAM-2686. One of attached to jira issue diagram can be used to test this change (see comment in jira issue). 

This is first part of the RHPAM-2686, small fix for missing SimData. Now marshallers doesn't fail with NullPointer exception but just ignore missing SimData.

hasys commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin, @LuboTerifaj,

I tested Full downstream and it is possible to load now, so the issue I was fixing is fixed, but it is not possible to save the process due to:
![image](https://user-images.githubusercontent.com/1477262/74311160-13f8c300-4d6f-11ea-80d6-478b03198b9a.png)

It looks like the issue is in annotations, and as temporary solution we will have to ask to remove annotations as well before it will be usable, which is still easier than remodel entire diagrams.

Sonar Cloud failing due to duplications in marshallers.

@romartin, @LuboTerifaj WDYT does it make sense to merge this PR?

romartin commented:
hey @hasys 

Sorry for late response. What's the current status for this? I mean, the bug when "saving" is introduced by this PR? Is it still something missing or can we do the review?

Thx!

hasys commented:
Jenkins execute full downstream build

hasys commented:
Hi @romartin, @LuboTerifaj,

I moved this fix to separate Jira issue and reworded the commit message and rebased to latest master. Now it should be easier to check/test/verify even if initial RHPAM-2686 won't be fully done before the release.

This PR is ready for review.

Thank you!

hasys commented:
Jenkins execute full downstream build

hasys commented:
Sonar Cloud failing due to duplications in marshallers.

<img width="611" alt="image" src="https://user-images.githubusercontent.com/1477262/74922191-26a67400-53cf-11ea-98c8-8e7da95e1579.png">


Pr. is ready for review.

LuboTerifaj commented:
Jenkins execute full downstream build

hasys commented:
Hi @jesuino, those test is unstable, can we disable it for now and report a Jira issue for stabilizations?

<details>
  <summary>org.kie.bc.DeploymentIT.org.kie.bc.DeploymentIT</summary>
  <pre>Error Message

Cannot deploy business-monitoring-webapp.war: {"WFLYCTL0062: Composite operation failed and was rolled back. Steps that failed:" => {"Operation step-1" => {"WFLYCTL0080: Failed services" => {"jboss.deployment.unit.\"business-monitoring-webapp.war\".WeldStartService" => "Failed to start service
    Caused by: org.jboss.weld.exceptions.DeploymentException: No filesystem for uri (git://system/system) found.
    Caused by: org.uberfire.java.nio.file.FileSystemNotFoundException: No filesystem for uri (git://system/system) found."}}}}

Stacktrace

org.jboss.arquillian.container.spi.client.container.DeploymentException: 
Cannot deploy business-monitoring-webapp.war: {"WFLYCTL0062: Composite operation failed and was rolled back. Steps that failed:" => {"Operation step-1" => {"WFLYCTL0080: Failed services" => {"jboss.deployment.unit.\"business-monitoring-webapp.war\".WeldStartService" => "Failed to start service
    Caused by: org.jboss.weld.exceptions.DeploymentException: No filesystem for uri (git://system/system) found.
    Caused by: org.uberfire.java.nio.file.FileSystemNotFoundException: No filesystem for uri (git://system/system) found."}}}}
</pre>
</details>

<details>
  <summary>org.kie.wb.test.rest.functional.ProjectIntegrationTest.testCreateProjectWithDescription</summary>
  <pre>Error Message

SERVER_ERROR: Request failed because of FileSystemNotFoundException: No filesystem for uri (default://master@projectTestSpace/projectToBeDeleted/) found.

Stacktrace

org.kie.wb.test.rest.client.ClientRequestTimedOutException: SERVER_ERROR: Request failed because of FileSystemNotFoundException: No filesystem for uri (default://master@projectTestSpace/projectToBeDeleted/) found.
	at org.kie.wb.test.rest.functional.ProjectIntegrationTest.createProject(ProjectIntegrationTest.java:396)
	at org.kie.wb.test.rest.functional.ProjectIntegrationTest.testCreateProjectWithDescription(ProjectIntegrationTest.java:85)

Standard Output

 >>> testCreateProjectWithDescription <<< 
10:30:59.298 [main] INFO  o.k.w.t.r.client.RestWorkbenchClient.createProject:179 - Creating project 'projectWithDescription' in space 'projectTestSpace'
10:30:59.431 [main] INFO  o.k.w.t.r.client.RestWorkbenchClient.getJob:118 - Getting job '1583163059430-101'
10:31:00.436 [main] INFO  o.k.w.t.r.client.RestWorkbenchClient.getJob:118 - Getting job '1583163059430-101'
10:31:01.439 [main] INFO  o.k.w.t.r.client.RestWorkbenchClient.getJob:118 - Getting job '1583163059430-101'
10:31:01.684 [main] WARN  o.k.w.t.r.client.RestWorkbenchClient.waitUntilJobFinished:466 -   Timeout waiting 2 seconds for job to succeed

</pre>
</details>

Hi @LuboTerifaj Full downstream build failed due to unrelated unstable test. Binaries are generated and this PR is ready for review. Thank you!

LuboTerifaj commented:
@hasys please see my comment in https://issues.redhat.com/browse/RHPAM-2727. Thanks!

jesuino commented:
Hello @hasys 

The DeploymentIT will be disabled until we fix what is wrong with it. I will be using this issue:

https://issues.redhat.com/browse/JBPM-8755

The problem is that this test never fails locally, I already tried a couple of times, it only fails on CI server.

Is the test `org.kie.wb.test.rest.functional.ProjectIntegrationTest.testCreateProjectWithDescription` unstable as well? I never saw a failure with it until now.

hasys commented:
> Hello @hasys
> 
> The DeploymentIT will be disabled until we fix what is wrong with it. I will be using this issue:
> 
> https://issues.redhat.com/browse/JBPM-8755
> 
> The problem is that this test never fails locally, I already tried a couple of times, it only fails on CI server.
> 
> Is the test `org.kie.wb.test.rest.functional.ProjectIntegrationTest.testCreateProjectWithDescription` unstable as well? I never saw a failure with it until now.

HI @jesuino, great, thank you! About the second test: at least it failed this time, you can see it in Full Downstream build or check stacktrace in my previous message.

hasys commented:
Hi @LuboTerifaj,

yes, it is still usable, because once you open it you can fix it and go further, no need to remodel the entire process.

LuboTerifaj commented:
Hey @hasys ,

One more note. All Business Rule Tasks have set value of Rule language property to null. When you change it and try to save the process, an exception is thrown:
<details><summary>java.lang.NullPointerException</summary>
<p>
ERROR [org.kie.workbench.common.stunner.project.backend.service.ProjectDiagramServiceController] (default task-4) Error while saving diagram with UUID [test].: java.lang.NullPointerException
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.properties.BusinessRuleTaskPropertyWriter.setRuleFlowGroup(BusinessRuleTaskPropertyWriter.java:58)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.tasks.TaskConverter.businessRuleTask(TaskConverter.java:202)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$when$4(NodeMatch.java:85)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch$Case.match(NodeMatch.java:140)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$apply$5(NodeMatch.java:113)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.LinkedList$LLSpliterator.tryAdvance(LinkedList.java:1249)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:499)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:486)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:531)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.apply(NodeMatch.java:115)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.tasks.TaskConverter.toFlowElement(TaskConverter.java:64)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$when$4(NodeMatch.java:85)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch$Case.match(NodeMatch.java:140)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.lambda$apply$5(NodeMatch.java:113)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.LinkedList$LLSpliterator.tryAdvance(LinkedList.java:1249)
	at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
	at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:499)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:486)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:531)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.NodeMatch.apply(NodeMatch.java:115)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.FlowElementConverter.toFlowElement(FlowElementConverter.java:55)
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:193)
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:175)
	at java.util.HashMap$ValueSpliterator.forEachRemaining(HashMap.java:1628)
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:482)
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:472)
	at java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:150)
	at java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:173)
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:485)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.ProcessConverterDelegate.convertChildNodes(ProcessConverterDelegate.java:71)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.processes.RootProcessConverter.convertProcess(RootProcessConverter.java:52)
	at org.kie.workbench.common.stunner.bpmn.backend.converters.fromstunner.DefinitionsConverter.toDefinitions(DefinitionsConverter.java:54)
	at org.kie.workbench.common.stunner.bpmn.backend.BaseDirectDiagramMarshaller.marshall(BaseDirectDiagramMarshaller.java:123)
	at org.kie.workbench.common.stunner.core.backend.service.AbstractVFSDiagramService.serialize(AbstractVFSDiagramService.java:262)
	at org.kie.workbench.common.stunner.project.backend.service.ProjectDiagramServiceController.save(ProjectDiagramServiceController.java:171)
	at org.kie.workbench.common.stunner.project.backend.service.ProjectDiagramServiceImpl.save(ProjectDiagramServiceImpl.java:158)
	at org.kie.workbench.common.stunner.project.backend.service.ProjectDiagramServiceImpl.save(ProjectDiagramServiceImpl.java:50)
	at org.kie.workbench.common.stunner.project.backend.service.ProjectDiagramServiceImpl$Proxy$_$$_WeldClientProxy.save(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.jboss.errai.bus.server.io.AbstractRPCMethodCallback.invokeMethodFromMessage(AbstractRPCMethodCallback.java:65)
	at org.jboss.errai.bus.server.io.ValueReplyRPCEndpointCallback.callback(ValueReplyRPCEndpointCallback.java:40)
	at org.jboss.errai.bus.server.io.RemoteServiceCallback.callback(RemoteServiceCallback.java:54)
	at org.jboss.errai.cdi.server.CDIExtensionPoints$2.callback(CDIExtensionPoints.java:448)
	at org.jboss.errai.bus.server.DeliveryPlan.deliver(DeliveryPlan.java:47)
	at org.jboss.errai.bus.server.ServerMessageBusImpl.sendGlobal(ServerMessageBusImpl.java:297)
	at org.jboss.errai.bus.server.SimpleDispatcher.dispatchGlobal(SimpleDispatcher.java:46)
	at org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:96)
	at org.jboss.errai.bus.server.service.ErraiServiceImpl.store(ErraiServiceImpl.java:113)
	at org.jboss.errai.bus.server.servlet.DefaultBlockingServlet.doPost(DefaultBlockingServlet.java:144)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:706)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:791)
	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:74)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:129)
	at io.undertow.websockets.jsr.JsrWebSocketFilter.doFilter(JsrWebSocketFilter.java:173)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.uberfire.ext.security.server.SecureHeadersFilter.doFilter(SecureHeadersFilter.java:110)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at org.uberfire.ext.security.server.SecurityIntegrationFilter.doFilter(SecurityIntegrationFilter.java:70)
	at io.undertow.servlet.core.ManagedFilter.doFilter(ManagedFilter.java:61)
	at io.undertow.servlet.handlers.FilterHandler$FilterChainImpl.doFilter(FilterHandler.java:131)
	at io.undertow.servlet.handlers.FilterHandler.handleRequest(FilterHandler.java:84)
	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
	at io.undertow.servlet.handlers.ServletChain$1.handleRequest(ServletChain.java:68)
	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:132)
	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
	at io.undertow.server.handlers.DisableCacheHandler.handleRequest(DisableCacheHandler.java:33)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.security.handlers.AuthenticationConstraintHandler.handleRequest(AuthenticationConstraintHandler.java:53)
	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
	at io.undertow.servlet.handlers.security.ServletSecurityConstraintHandler.handleRequest(ServletSecurityConstraintHandler.java:59)
	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at org.wildfly.extension.undertow.deployment.GlobalRequestControllerHandler.handleRequest(GlobalRequestControllerHandler.java:68)
	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
	at org.wildfly.extension.undertow.security.SecurityContextThreadSetupAction.lambda$create$0(SecurityContextThreadSetupAction.java:105)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at org.wildfly.extension.undertow.deployment.UndertowDeploymentInfoService$UndertowThreadSetupAction.lambda$create$0(UndertowDeploymentInfoService.java:1502)
	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:360)
	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:830)
	at org.jboss.threads.ContextClassLoaderSavingRunnable.run(ContextClassLoaderSavingRunnable.java:35)
	at org.jboss.threads.EnhancedQueueExecutor.safeRun(EnhancedQueueExecutor.java:1985)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.doRunTask(EnhancedQueueExecutor.java:1487)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1378)
	at java.lang.Thread.run(Thread.java:748)
</p>
</details>

Thanks!

hasys commented:
@LuboTerifaj,

great catch! Pleas report it as an addition point or subjira of [RHPAM-2686](https://issues.redhat.com/browse/RHPAM-2686). Thank you!

LuboTerifaj commented:
Hello @hasys , how you can fix the issues I mentioned above? If the process is broken, xml should be open instead.

hasys commented:
Hi @LuboTerifaj,

you can create new Ruleflow task, copy/paste old node and connect new instead. After it you can save the process.

Anyway it is not part of this jira issue, and not part of the acceptance criteria. The goal of this PR is to open process with empty BPSimData.

Validation and checking all possible issues in all possible nodes not part of this jira issue, this task is part of [RHPAM-2686](https://issues.redhat.com/browse/RHPAM-2686).

Thank you!

hasys commented:
Thank you @LuboTerifaj!

@romartin this PR is ready to merge, Full downstream build failed tests are unrelated, see https://github.com/kiegroup/kie-wb-common/pull/3153#issuecomment-594513421

Thank you!
